File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
  def largestRectangleArea(self, heights: List[int]) -> int:
    ans = 0
    stack = [
      {
        "i": 0,
        "s": 0,
        "h": heights[0],
      }
    ]
    popped = None
    for i in range(1, len(heights)):
      if heights[i] <= stack[-1]["h"]:
        popped = stack.pop()
        ans = max(ans, (popped["i"]-popped["s"]+1)*popped["h"])
        while stack and stack[-1]["h"] >= heights[i]:
          stack[-1]["i"] = popped["i"]
          popped = stack.pop()
          ans = max(ans, (popped["i"]-popped["s"]+1)*popped["h"])
        stack.append({
          "i": i,
          "s": popped["s"],
          "h": heights[i],
        })
      else:
        stack.append({
          "i": i,
          "s": i,
          "h": heights[i],
        })
    while stack:
      popped = stack.pop()
      ans = max(ans, (popped["i"]-popped["s"]+1)*popped["h"])
      if stack:
        stack[-1]["i"] = popped["i"]
    return ans

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.rb
# @param {Integer[]} heights
# @return {Integer}
def largest_rectangle_area(heights)
  ans = 0
  stack = [
    {
      i: 0,
      s: 0,
      h: heights[0],
    }
  ]
  popped = nil
  (1...heights.length).each do |i|
    if heights[i] <= stack[-1][:h]
      popped = stack.pop
      ans = [ans, (popped[:i]-popped[:s]+1)*popped[:h]].max
      while !stack.empty? && stack[-1][:h] >= heights[i]
        stack[-1][:i] = popped[:i]
        popped = stack.pop
        ans = [ans, (popped[:i]-popped[:s]+1)*popped[:h]].max
      end
      stack << {
        i: i,
        s: popped[:s],
        h: heights[i],
      }
    else
      stack << {
        i: i,
        s: i,
        h: heights[i],
      }
    end
  end
  while !stack.empty?
    popped = stack.pop
    ans = [ans, (popped[:i]-popped[:s]+1)*popped[:h]].max
    stack[-1][:i] = popped[:i] if !stack.empty?
  end
  ans
end